src/Language/Parser/Expression/ExpressionParser.php

```diff
@@ -0,0 +1,576 @@
+                TokenType::PERIOD,
+                TokenType::OPTCHAIN => $this->parseAcccess($tokens, $result),
+                default => $result
+            };
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param \Iterator<mixed,Token> $tokens
+     * @return ExpressionNode
+     */
+    private function parseUnaryOperation(\Iterator &$tokens): ExpressionNode
+    {
+        $startingToken = $tokens->current();
+
+        $operator = $this->parseUnaryOperator($tokens);
+        $operand = $this->parseUnaryStatement($tokens);
+
+        $unaryOperationNode = new UnaryOperationNode(
+            rangeInSource: Range::from(
+                $startingToken->boundaries->start,
+                $operand->rangeInSource->end
+            ),
+            operator: $operator,
+            operand: $operand
+        );
+
+        return new ExpressionNode(
+            rangeInSource: $unaryOperationNode->rangeInSource,
+            root: $unaryOperationNode
+        );
+    }
+
+    /**
+     * @param \Iterator<mixed,Token> $tokens
+     * @return UnaryOperator
+     */
+    private function parseUnaryOperator(\Iterator &$tokens): UnaryOperator
+    {
+        $unaryOperator = match (Scanner::type($tokens)) {
+            TokenType::OPERATOR_BOOLEAN_NOT => UnaryOperator::NOT,
+            default => throw ExpressionCouldNotBeParsed::becauseOfUnexpectedToken(
+                expectedTokenTypes: TokenTypes::from(TokenType::OPERATOR_BOOLEAN_NOT),
+                actualToken: $tokens->current()
+            )
+        };
+
+        Scanner::skipOne($tokens);
+
+        return $unaryOperator;
+    }
+
+    private function withStopAt(TokenType $stopAt): self
+    {
+        $newExpressionParser = clone $this;
+        $newExpressionParser->stopAt = $stopAt;
+
+        return $newExpressionParser;
+    }
+
+    private function withPrecedence(Precedence $precedence): self
+    {
+        $newExpressionParser = clone $this;
+        $newExpressionParser->precedence = $precedence;
+
+        return $newExpressionParser;
+    }
+
+    /**
+     * @param \Iterator<mixed,Token> $tokens
+     * @return boolean
+     */
+    private function shouldStop(\Iterator &$tokens): bool
+    {
+        Scanner::skipSpaceAndComments($tokens);
+
+        if (Scanner::isEnd($tokens)) {
+            return true;
+        }
+
+        $type = Scanner::type($tokens);
+
+        if ($this->precedence->mustStopAt($type)) {
+            return true;
+        }
+
+        if ($this->stopAt && $type === $this->stopAt) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param \Iterator<mixed,Token> $tokens
+     * @return ExpressionNode
+     */
+    private function parseBooleanLiteral(\Iterator &$tokens): ExpressionNode
+    {
+        $this->booleanLiteralParser ??= BooleanLiteralParser::singleton();
+
+        $booleanLiteralNode = $this->booleanLiteralParser->parse($tokens);
```

Comment:
I dont think, we need this second instance cache layer. Calling `::singleton()` every-time should be fine?

Also it seems we are not using it for tests to swap out implementations...

Reply:
The underlying consideration is:
```
using($this->booleanLiteralParser) = cost_of(ReferenceLookup)
vs.
using(BooleanLiteralParser::singleton()) = cost_of(FunctionCall) + cost_of(ReferenceLookup)
```

It's just a performance optimization (maybe a bit over the top, but it's not nothing :D).

Reply:
As I'm working on #34, I noticed that I'm just one step away from being able to make all parsers static.

That will effectively eliminate this problem :)